### PR TITLE
Prepare for client steering

### PR DESCRIPTION
**Pull request by RockyMandayam2**
Prepare for client steering (hooks to actually steer up/down or deauthenticate are not done yet). Lots of TODOs left (this is just the most basic incomplete logic).

I tested just the fact that I can send an HTTP put for the ClientSteering algorithm and it returns an empty map (no clients currently connected to APs on our gateway instance), but it enters the `computeApClientActionMap` method.

### CODE REVIEW

File: owrrm/src/main/java/com/facebook/openwifi/rrm/RRMAlgorithm.java [hunk 1]
```diff
@@ -72,6 +78,12 @@ public static class AlgorithmResult {
 		 * @see TPC#computeTxPowerMap()
 		 */
 		public Map<String, Map<String, Integer>> txPowerMap;
+		/**
```
Comment (elludraon) [inline owrrm/src/main/java/com/facebook/openwifi/rrm/RRMAlgorithm.java:81]: nit: add blank line before

File: owrrm/src/main/java/com/facebook/openwifi/rrm/RRMAlgorithm.java [hunk 2]
```diff
@@ -183,7 +195,7 @@ public AlgorithmResult run(
 			ChannelOptimizer optimizer;
 			switch (mode) {
 			default:
-				if (!allowDefaultMode || !mode.isEmpty()) {
```
Comment (elludraon) [inline owrrm/src/main/java/com/facebook/openwifi/rrm/RRMAlgorithm.java:186]: this was correct before, why did you change it? (error if either default/empty is forbidden OR invalid non-empty mode name was passed)

File: owrrm/src/main/java/com/facebook/openwifi/rrm/RRMAlgorithm.java [hunk 3]
```diff
@@ -286,6 +298,40 @@ public AlgorithmResult run(
...
+				// fall through
+			case SingleAPBandSteering.ALGORITHM_ID:
+				optimizer = SingleAPBandSteering.makeWithArgs(
+					modeler.getDataModelCopy(),
+					zone,
+					deviceDataManager,
+					args
+				);
+				break;
+
```
Comment (elludraon) [inline owrrm/src/main/java/com/facebook/openwifi/rrm/RRMAlgorithm.java:326]: nit: delete extraneous line

File: owrrm/src/main/java/com/facebook/openwifi/rrm/optimizers/clientsteering/ClientSteeringOptimizer.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+import java.util.Map;
+
+import com.facebook.openwifi.rrm.DeviceConfig;
+import com.facebook.openwifi.rrm.DeviceDataManager;
+import com.facebook.openwifi.rrm.modules.Modeler.DataModel;
+
+/** Client steering base class */
+public abstract class ClientSteeringOptimizer {
+
+	/** Represents the client steering related actions an AP can take */
```
Comment (elludraon) [inline owrrm/src/main/java/com/facebook/openwifi/rrm/optimizers/clientsteering/ClientSteeringOptimizer.java:20]: - javadoc only applies to the first element, so the second two variables will get nothing -- either add to all or none
- do you want to make an enum type instead?

File: owrrm/src/main/java/com/facebook/openwifi/rrm/optimizers/clientsteering/ClientSteeringOptimizer.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+	/** The input data model. */
+	protected final DataModel model;
+
+	/** The RF zone. */
+	protected final String zone;
+
+	/** The device configs within {@link #zone}, keyed on serial number. */
+	protected final Map<String, DeviceConfig> deviceConfigs;
+
+	public ClientSteeringOptimizer(
```
Comment (elludraon) [inline owrrm/src/main/java/com/facebook/openwifi/rrm/optimizers/clientsteering/ClientSteeringOptimizer.java:34]: add javadoc ("Constructor" is fine)

File: owrrm/src/main/java/com/facebook/openwifi/rrm/optimizers/clientsteering/ClientSteeringOptimizer.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+import com.facebook.openwifi.rrm.DeviceConfig;
+import com.facebook.openwifi.rrm.DeviceDataManager;
+import com.facebook.openwifi.rrm.modules.Modeler.DataModel;
+
+/** Client steering base class */
+public abstract class ClientSteeringOptimizer {
+
+	/** Represents the client steering related actions an AP can take */
+	public static final String STEER_UP_ACTION = "STEER_UP";
+	public static final String STEER_DOWN_ACTION = "STEER_DOWN";
```
Comment (elludraon) [inline owrrm/src/main/java/com/facebook/openwifi/rrm/optimizers/clientsteering/ClientSteeringOptimizer.java:22]: you may want to think about implementing in a way that will support more than 2 bands?

File: owrrm/src/main/java/com/facebook/openwifi/rrm/optimizers/clientsteering/ClientSteeringOptimizer.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+	}
+
+	/**
+	 * Computed map from AP serial number to client MAC to client steering
+	 * action.
+	 */
+	public abstract Map<String, Map<String, String>> computeApClientActionMap();
+
+	/**
+	 * Actually steer clients (steer up, steer down, and deauthenticate).
```
Comment (elludraon) [inline owrrm/src/main/java/com/facebook/openwifi/rrm/optimizers/clientsteering/ClientSteeringOptimizer.java:61]: nit: delete "actually"

File: owrrm/src/main/java/com/facebook/openwifi/rrm/optimizers/clientsteering/SingleAPBandSteering.java [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+
+import com.facebook.openwifi.cloudsdk.UCentralConstants;
+import com.facebook.openwifi.cloudsdk.UCentralUtils;
+import com.facebook.openwifi.cloudsdk.models.ap.Capabilities;
+import com.facebook.openwifi.cloudsdk.models.ap.State;
+import com.facebook.openwifi.rrm.DeviceDataManager;
+import com.facebook.openwifi.rrm.modules.Modeler.DataModel;
+import com.facebook.openwifi.rrm.modules.ModelerUtils;
+
+public class SingleAPBandSteering extends ClientSteeringOptimizer {
```
Comment (elludraon) [inline owrrm/src/main/java/com/facebook/openwifi/rrm/optimizers/clientsteering/SingleAPBandSteering.java:23]: add javadoc

File: owrrm/src/main/java/com/facebook/openwifi/rrm/optimizers/clientsteering/SingleAPBandSteering.java [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+	private static final short DEFAULT_MAX_RSSI_2G = -67;
+	/**
+	 * RSSI (dBm) below which a client on 5G/6G should be requested to move to
+	 * 2G
+	 */
+	private static final short DEFAULT_MIN_RSSI_NON_2G = -82;
+
+	private final short minRssi2G;
+	private final short maxRssi2G;
+	private final short minRssiNon2G;
```
Comment (elludraon) [inline owrrm/src/main/java/com/facebook/openwifi/rrm/optimizers/clientsteering/SingleAPBandSteering.java:49]: add javadoc

File: owrrm/src/main/java/com/facebook/openwifi/rrm/optimizers/clientsteering/SingleAPBandSteering.java [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+						capabilitiesPhy
+					);
+					if (band == null) {
+						continue;
+					}
+					for (
+						State.Interface.SSID.Association assoc : ssid.associations
+					) {
+						// TODO 11k beacons: check which bands client can use
+						// TODO check which bands AP can use
```
Comment (elludraon) [inline owrrm/src/main/java/com/facebook/openwifi/rrm/optimizers/clientsteering/SingleAPBandSteering.java:148]: duplicate TODOs

File: owrrm/src/main/java/com/facebook/openwifi/rrm/RRMAlgorithm.java [hunk 11]
```diff
@@ -42,6 +44,10 @@ public enum AlgorithmType {
 		OptimizeTxPower (
 			"Optimize tx power configuration",
 			"Run transmit power control algorithm"
+		),
+		ClientSteering (
+			"Optimize AP and band for each client",
```
Comment (elludraon) [inline owrrm/src/main/java/com/facebook/openwifi/rrm/RRMAlgorithm.java:49]: "Steer clients onto the optimal AP and band"

**Review by elludraon — CHANGES_REQUESTED** [review]

File: owrrm/src/main/java/com/facebook/openwifi/rrm/RRMAlgorithm.java (hunk 2 above)
Comment (RockyMandayam2) [reply owrrm/src/main/java/com/facebook/openwifi/rrm/RRMAlgorithm.java:186]: I misunderstood something, let me change it back.

File: owrrm/src/main/java/com/facebook/openwifi/rrm/optimizers/clientsteering/ClientSteeringOptimizer.java [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+		String apSerialNumber,
+		String bssid
+	) {
+		Map<String, Long> clientLatestAttempt = apRadioLastAttempt
+			.get(apSerialNumber);
+		if (clientLatestAttempt == null) {
+			return Long.MIN_VALUE;
+		}
+		return clientLatestAttempt.getOrDefault(bssid, Long.MIN_VALUE);
+	}
```
Comment (elludraon) [inline owrrm/src/main/java/com/facebook/openwifi/rrm/optimizers/clientsteering/ClientSteeringOptimizer.java:69]: This doesn't belong here. To make this cleaner, I would put this into a new class like `ClientSteeringState` which can act as a singleton (i.e., `static ClientSteeringState getInstance() { return INSTANCE; }`, and these methods should be non-static). Then you can pass the single instance of `ClientSteeringState` into `ClientSteeringOptimizer` in the constructor.

File: owrrm/src/main/java/com/facebook/openwifi/rrm/optimizers/clientsteering/ClientSteeringOptimizer.java [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+			.computeIfAbsent(apSerialNumber, k -> new TreeMap<>())
+			.put(bssid, System.currentTimeMillis());
+	}
+
+	/**
+	 * Get the time of the latest client steering attempt by AP serial number
+	 * and radio bssid. Return {@link Long#MIN_VALUE} if no client steering
+	 * attempt has been made for the given AP and radio.
+	 */
+	protected static final long getLatestClientSteeringAttempt(
```
Comment (elludraon) [inline owrrm/src/main/java/com/facebook/openwifi/rrm/optimizers/clientsteering/ClientSteeringOptimizer.java:59]: just return boxed `Long` so you can use null? don't use `Long.MIN_VALUE`, that doesn't make sense

File: owrrm/src/main/java/com/facebook/openwifi/rrm/optimizers/clientsteering/ClientSteeringOptimizer.java [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+	 * Register the time of the latest client steering attempt by AP serial
+	 * number and radio bssid.
+	 */
+	protected static final void registerClientSteeringAttempt(
+		String apSerialNumber,
+		String bssid
+	) {
+		apRadioLastAttempt
+			.computeIfAbsent(apSerialNumber, k -> new TreeMap<>())
+			.put(bssid, System.currentTimeMillis());
```
Comment (elludraon) [inline owrrm/src/main/java/com/facebook/openwifi/rrm/optimizers/clientsteering/ClientSteeringOptimizer.java:51]: Use monotonic time, i.e. `System.nanoTime()` especially when the epoch is not relevant

File: owrrm/src/main/java/com/facebook/openwifi/rrm/optimizers/clientsteering/ClientSteeringOptimizer.java [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+	}
+
+	/**
+	 * Map from AP serial number to radio bssid to time (unix time in ms) of the
+	 * latest attempted client steering action. It is static here to track
+	 * attempts across all instances of all subclasses. The {@code Long} values
+	 * are never null.
+	 */
+	private static final Map<String, Map<String, Long>> apRadioLastAttempt =
+		new TreeMap<>();
```
Comment (elludraon) [inline owrrm/src/main/java/com/facebook/openwifi/rrm/optimizers/clientsteering/ClientSteeringOptimizer.java:39]: Why `TreeMap` here? Also, if this is static then you need to use a concurrent structure like `ConcurrentHashMap`.
